File: src/main/java/ting/leetcode/java/binarytree/MaximumDepthOfBinaryTree.java
package ting.leetcode.java.binarytree;

public class MaximumDepthOfBinaryTree {
    int maxDepth = 0;
    int currentDepth = 0;
    public int maxDepth(TreeNode root) {
        traverse(root);
        return maxDepth;
    }

    private void traverse(TreeNode node) {
        if (node == null) {
            return;
        }

        // 前序位置
        currentDepth++;

        if (node.left == null && node.right == null) {
            maxDepth = Math.max(maxDepth, currentDepth);
        }

        traverse(node.left);
        // 中序位置
        traverse(node.right);
        // 后序位置
        currentDepth--;
    }


    int maxDepthRecursion(TreeNode root) {
        if (root == null) {
            return 0;
        }
        // 利用定义，计算左右子树的最大深度
        int leftMax = maxDepthRecursion(root.left);
        int rightMax = maxDepthRecursion(root.right);
        // 整棵树的最大深度等于左右子树的最大深度取最大值，
        // 然后再加上根节点自己

        return Math.max(leftMax, rightMax) + 1;
    }
}

File: src/main/kotlin/ting/leetcode/tree/MaximumDepthOfBinaryTree.kt
package ting.leetcode.tree

import kotlin.math.max

class MaximumDepthOfBinaryTree {
    fun maxDepth(root: TreeNode?): Int {
        if (root == null) return 0
        val result = mutableListOf<MutableList<TreeNode?>>();
        val queue = ArrayDeque<TreeNode?>();
        queue.addLast(root)

        while (queue.isNotEmpty()) {
            val list = mutableListOf<TreeNode?>()
            for (i in queue.indices) {
                val node = queue.removeFirst()
                list.add(node)
                if (node?.left != null) queue.addLast(node.left)
                if (node?.right != null) queue.addLast(node.right)
            }
            result.add(list);
        }
        return result.size
    }

    fun maxDepthRecursion(root: TreeNode?): Int =
        if (root == null) 0 else max(maxDepthRecursion(root.left), maxDepthRecursion(root.right)) + 1
}